odoo/addons/base/models/res_company.py

```diff
@@ -376,24 +376,74 @@ def _get_main_company(self):
     def __accessible_branches(self):
         # Get branches of this company that the current user can use
         self.ensure_one()
-        children = self.search([('id', 'child_of', self.id)])
+
+        rslt = []
+        current = self.sudo()
         accessible = self.env.companies
-        return (children & accessible).ids
+        while current:
+            rslt.extend((current & accessible).ids)
+            current = current.child_ids
+
+        if not rslt and self.env.su:
```

Comment:
Why only `if not rslt`?
I think the issue is only for `SUPERUSER_ID`, `env.su` is already managed in `env.companies`

Reply:
Note that `su` (or `uid`) is missing from the `ormcache` keys